### ISSUE TITLE
net_Disconnect bug?

### ISSUE THREAD

**Issue by toorisrael** [yourmnbbn/tiny-csgo-client#3]
Sometimes I've got disconnect messages, but account is not really disconnecting - no disconnect in server logs, account is still connected since a long time:
![image](https://user-images.githubusercontent.com/4301144/172224737-d2bb2b09-adaf-4b14-a912-0c770b5793cb.png)

**Comment by yourmnbbn** [comment]
This is actually related to #2. Server won't send disconnect message with empty text. Although the message parsing is taken from the leaked source, they must have changed something. So the message parsing is sometimes corrupted. I've simply checked the latest binary to see if there is any new messages added, here is what I have found.
```
CNetMessagePB<39, class CSVCMsg_HltvFixupOperatorStatus, 3, 1>
CNetMessagePB<21,class CCLCMsg_HltvFixupOperatorTick,0,1>
```
And it's also possible they add or remove some fields in the existing messages. Theoretically even if we have missed the new message handling or have trouble processing the new fields of the existing, the message stream contains the length of each sub message so we can simply bypass it. They actually have little impact on the parsing. But the client still got many wrong message type like 0x5C, which proves that our message parsing is outdated. That's why there are so many unexpected behavior during message parsing. 

This has to be fixed in the hard way, with the help of RE, which also takes much longer time.

**Comment by toorisrael** [comment]
I don't know if it's any helpful, but I've found [this](https://github.com/smefpw/smefs-Indigo-Remastered/blob/master/INDIGO/protobuf/netmessages.pb.h#L1372) - it does include mentioned 2 new messages, so it may be more up to date.

Also, we've talked about the memory before. This is run without Sandboxie:
![image](https://user-images.githubusercontent.com/4301144/172389158-4cb8a57e-da04-4470-b6d6-d43efd57c3d7.png)
In Sandboxie it behaves the same, allocating 280M and using only ~10M per process (in this case it's 40M). Is it 100% normal?

**Comment by yourmnbbn** [comment]
> I don't know if it's any helpful, but I've found [this](https://github.com/smefpw/smefs-Indigo-Remastered/blob/master/INDIGO/protobuf/netmessages.pb.h#L1372) - it does include mentioned 2 new messages, so it may be more up to date.

It's not about the missing message or changed message, it's about message stream parsing. I've figured out why. Some of the message is still encoded after the raw packet data decryption. Message type 0x5C is actually 0x17, which is svc_UserMessage. After some math I've written a decode function. After decoding, all the messages are now successfully parsed. But I'm still trying to find when would the messages being encoded, is there some flags hidden in the stream?
```cpp
uint8_t decode(uint8_t input)
{
    uint8_t mod = input % 4;
    return mod ? ((1 << (5 + mod)) + (input - mod) / 4) : (input / 4);
}
```

> Also, we've talked about the memory before. This is run without Sandboxie:
In Sandboxie it behaves the same, allocating 280M and using only ~10M per process (in this case it's 40M). Is it 100% normal?

Since I have never paid attention to the memory usage, I am not sure if it's normal. After doing some of my prior todos I will take a look at the memory issue.

**Comment by toorisrael** [comment]
All last changes are great, I'm really impressed. However it crashes on receiving chat message from server. The message contains special characters, idk if it's important:
![image](https://user-images.githubusercontent.com/4301144/172645947-2fd11812-dff4-426e-9828-9abdd2d949db.png)
Also I haven't checked code yet, maybe it's a problem with recognizing client who sends message, cause it's the sever?

**Comment by yourmnbbn** [comment]
I don’t know what cause the crash, it’s fine in my test case but the crash make sense because the TestMsg, SayText and SayText2 user messages are extremely unstable. And some parts of the string is broken, especially those next to the blank space.  I think it’s related to the decode I just added. The decoded messages larger than 63 bytes are not readable, because the size bits are broken, message can’t be read in correct size.  That’s why the client now still get unhandled message error. It’s weird that it only works with small packets. Maybe I’ve missed something.  I’m still working on this. 

Good news is that the decode at least helps to read small packet of messages so we can now reconnect when server changelevel. To me I think this is more important than those user messages.

**Comment by yourmnbbn** [comment]
@toorisrael When working on some other project I've found that the latest tier0.dll from csgo would allocate a huge amount of memory. So that's the reason of the memory issue. I use the one I've compiled using the leaked source and the memory usage dropped from ~320M to ~60M. You can use it to have a try or compile it on your own. And maybe you can try to find the problems of the tier0 project to further reduce the memory usage if you really care about it.

EDIT: Latest commit has updated the tier0.dll, I've recompiled it using the much newer c++ standard and compiler.

**Comment by yourmnbbn** [comment]
This issue is related to #2, which has been fixed by commit 80130cd.
